crates/chain/src/chain_data.rs

```diff
@@ -215,4 +241,64 @@ impl<I: ChainPosition> FullTxOut<I> {
     }
 }
 
+impl<A: Anchor> FullTxOut<ObservedAs<A>> {
+    /// Whether the `txout` is considered mature.
+    ///
+    /// This is the alternative version of [`is_mature`] which depends on `chain_position` being a
+    /// [`ObservedAs<A>`] where `A` implements [`Anchor`].
+    ///
+    /// [`is_mature`]: Self::is_mature
+    pub fn is_mature(&self, tip: u32) -> bool {
+        if !self.is_on_coinbase {
+            return true;
+        }
+
+        let tx_height = match &self.chain_position {
+            ObservedAs::Confirmed(anchor) => anchor.confirmation_height_upper_bound(),
```

Comment:
This seems confusing to me. It will return false for a coinbase tx which has it's anchor block at "less than 100 depth" from chain tip, while the actual transaction could be deep enough to be called "matured". 

corollary question: Should we even try to calculate maturity with Anchors? Or maybe create a default anchor that has more metadata (tx confirmation height) than `BlockId`? 

We can only have the situation: "if anchor is mature, then the transaction is mature too". But it can very well be the other way around. I.e., The transaction is matured but the anchor isn't. This will give false negatives. 

If we are happy with this, it should be explained in the doc comments.

Reply:
False negatives are fine imo. If the `Anchor` implementation has the exact confirmation height, `confirmation_height_upper_bound()` will give the actual confirmation height.

Docs are updated!

Reply:
> False negatives are fine imo.

I don't see how that won't screw up wallet operation for coinbase totally. False positive will create invalid tx, while false negetive would miss the correct spendable balance. But okay. Can be fixed later.

Reply:
@LLFourn do you think we should introduce an extension trait for `Anchor`? Originally, I had the `ConfirmationHeight` trait.

Reply:
> @LLFourn do you think we should introduce an extension trait for Anchor? Originally, I had the ConfirmationHeight trait.

I tend to fall toward this also, which I think was there in a previous version. Corollary question, if we end up putting confirmation height in anchor anyway, can't we just call Anchors as the transaction inclusion Block Height and Block Hash? Then `BlockId` could be the "minimal anchor". And extra metadata could be added later as user's choice.

Reply:
I am confused by the discussion here. @evanlinjin  pointed out that false negatives are impossible but @rajarshimaitra is concerned about false positives which can't occur and then @evanlinjin :+1: this post. What problem are we solving by the `ConfirmationHeight` trait?

Reply:
The gist of the issue is maturity calculation would require knowledge of the exact confirmation height. By using the anchor's height we will undershoot maturity. I.e., when a coinbase transaction is confirmed at block 50, anchor is at 100, calculating maturity at tip 150 would give "immature" whereas the transaction, in reality, got matured. 

Same case for timelocked txs. So we have to know and store the confirmation height somewhere. Using the anchor's height here would give the wrong result. That was my point.

Reply:
>  @evanlinjin pointed out that false negatives are impossible but @rajarshimaitra is concerned about false positives which can't occur

to clarify, 
`false positive`: calling an immature, mature. This won't occur in this case because the anchor can at minimum be the inclusion block. 
`false negative`: calling a mature, immature. This occurs in this case because anchors can be descendants of the inclusion block.

Reply:
> > False negatives are fine imo.
> 
> I don't see how that won't screw up wallet operation for coinbase totally. False positive will create invalid tx, while false negetive would miss the correct spendable balance. But okay. Can be fixed later.

I think Lloyd is getting confused with this statement because I've implicitly mentioned that only false-negatives are possible.

Reply:
> The gist of the issue is maturity calculation would require knowledge of the exact confirmation height. By using the anchor's height we will undershoot maturity. I.e., when a coinbase transaction is confirmed at block 50, anchor is at 100, calculating maturity at tip 150 would give "immature" whereas the transaction, in reality, got matured.

but we don't know that it's confirmed at 50. We don't know when it was confirmed we only know an anchor for it so we give the best answer we can given the information we have. If your anchor type has confirmation height info then you will get the exactly correct answer. I don't think another trait helps the situation here unless you feel it is confusing that we allow you to call this even when we don't have exactly the correct answer and so you want constrain this method to that trait. Is this your concern @rajarshimaitra?

I'd say giving the best answer we can and document that `is_mature` can have false negatives depending on the `Anchor` implementation is best but I'm not too fussed here.

Reply:
Either I am missing something or I am not understanding the argument at all. The `is_mature` calculation with `Anchor` isn't just confusing, it's wrong. `is_mature` is used in balance calculation, so the calculated balance will also be wrong. How is this not a problem? 
 
 > If your anchor type has confirmation height info then you will get the exactly correct answer.
 
If the solution is another Anchor type that has the correct height, then let's just add it here, because this will be needed in bdk wallet too.

Reply:
> but we don't know that it's confirmed at 50. We don't know when it was confirmed we only know an anchor for it so we give the best answer we can given the information we have

We will always know the exact confirmation height of a tx at the time of syncing. We are just choosing not to store it and abstract it with an Anchor. Isn't this important information that any kind of tracker would need to track?

Reply:
Anyway, I don't wanna prolong this discussion further. To summarize it seems there are two possible ways to fix this problem for the wallet.

 - Define another anchor that has confirmation height in it. Use that in bdk wallet. 
 
 - Use Anchor BlockId as the inclusion block, and do not abstract it away with a descendant. Do not change the Anchor for a tx except for reorgs.
 
Both of these can be done later outside of this PR and maybe at another layer. So I am marking this resolved from my side unless others have any more input.